### ISSUE TITLE
Fine-tuning/Speaker adaptation

### ISSUE THREAD

**Issue by ZohaibAhmed** [tiberiu44/TTS-Cube#9]
First off, this is great work. Can't wait to play around with the code 👍 

In the training instructions, I see that you do have multispeaker support. Is it possible to "fine-tune" from an existing checkpoint with another dataset using `--resume`? Has anyone tried it and see if the results are good?

**Comment by tiberiu44** [comment]
I'll upload the models and instructions tonight. Out of curiosity, what data are you using for training?

**Comment by tiberiu44** [comment]
@ZohaibAhmed 

I have just uploaded some pre-trained models. In `data/models`, you have two folders: `en` and `ro`. For en you only have the vocoder (I'm working on an encoder that uses G2P, because the results were not ok with character-only features - see [https://github.com/tiberiu44/TTS-Cube/tree/master/examples/english/no-g2p](https://github.com/tiberiu44/TTS-Cube/tree/master/examples/english/no-g2p) for details).

The ro(manian) folder contains multi-speaker trained models for the encoder and decoder. Just copy the contents of `data/models/ro` to `data/models` and expand the `.gz` files. 

Than, you should be able to use the models with `synthesize.py`:

```bash
echo "Acesta este un simplu test." > input.txt
python3 cube/synthesis.py --input-file=input.txt --output-file=output.wav --speaker=anca --use-gpu
```

The list of speakers is: ANCA, BAS, CAU, DCS, DDM, EME, FDS, HTM, IPS, PCS, PMM, PSS, RMS, SAM, SDS, SGS, TIM, TSS. Use lowercase letters.

My guess is that you can use the multispeaker vocoder trained on RO for any other language. I've tried it on LJ and it sounded ok. However, the model in en/rnn_vocoder.network.gz is trained on LJ (single speaker).

Hope this helps.

**Comment by ZohaibAhmed** [comment]
Thanks @tiberiu44. I tried experimenting with a random audiobook from youtube (https://www.youtube.com/results?search_query=the+old+man+audiobook), I uploaded how a sample from the Vocoder here: https://drive.google.com/open?id=1hM9oAYjzC9qFcusxFijSMzdR_RbAugpQ

After your comments though, it would have made much more sense of training it on a cleaner dataset like LJSpeech and then fine-tuning it with this dataset. I'll try that this weekend.

**Comment by tiberiu44** [comment]
I just listened to your sample. I suppose audiobook version used for training is exactly this one: https://www.youtube.com/watch?v=FReY9fic00A. 

I think the speech quality will get much better. LJ is not than clean. It has some good quality and some low-quality samples. I think you should keep training this model and see what you can get from it. 

Is your batch-size 8000 or are you using a different value?

**Comment by tiberiu44** [comment]
I found nothing that would help. The loss has nothing to do with quality. The samples [here](https://tiberiu44.github.io/TTS-Cube/) (epoch 30) have the same loss as those from epoch 9, but they are of much higher quality. In fact, those from epoch 9 sound similar to the ones you shared. 

Just let the model train for one more week and check periodically. I would also love to see how it performs in time.

Also, if you find anything to help determine the quality automatically, it would be great to implement it. The models I uploaded are from one month of training. 

I'm currently working on sparsifying the network and implementing a runtime version of the system which takes advantage of this. I have an approved [DyNET PR](https://github.com/clab/dynet/pull/1437) with this. Theoretically, you would get real-time speech synthsis with this.

**Comment by ZohaibAhmed** [comment]
@tiberiu44 Checking in with an update, I continued to train the model for 3 more days, but didn't see much improvement: https://drive.google.com/open?id=1hM9oAYjzC9qFcusxFijSMzdR_RbAugpQ

I ran the encoder for another 2-3 days now, and its results sound OK (pronunciation is off, but that is probably just the data used for training): https://drive.google.com/open?id=12E6YjPdlusucju-93XskZg6_wDx0Jdts

![image](https://user-images.githubusercontent.com/660224/46514973-0bb17f00-c82f-11e8-98aa-5265d313b8dd.png)

I've also been trying to use https://github.com/fatchord/WaveRNN as the vocoder, and pass in the melspectrogram that your Encoder produced. The modified Vocoder he has seems to train very very quickly.

**Comment by tiberiu44** [comment]
@ZohaibAhmed, thanks for the update. 

I listened to the samples. I think you could still get better results with the vocoder. I've just updated the trainer to take `--learning-rate` as a parameter. If you still want to experiment, you could try resuming the training process on the Vocoder with a higher learning rate (0.001, instead of 0.0001 which was the default value).

Also, I think the Encoder can get much better. I didn't have the chance to fully integrate the G2P module. I'll try to do this during the weekend. After the update, you'll have to reimport the corpus and retrain the Encoder. The Vocoder will not be affected by this.

Let me know how you do with the WaveRNN vocoder. I'm curious on the difference in quality between the two approaches. I haven't got the chance to compare them in similar conditions.

Best,
Tibi

**Comment by tiberiu44** [comment]
Hi @ZohaibAhmed ,
I've just merged the branch with a new Vocoder. It is able to run faster than real-time: 
```
	3390/13095 processing file data/processed/train/LJ003-0336 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg loss=0.5805849801926386 execution time=0.8243229389190674
	3391/13095 processing file data/processed/train/LJ032-0044 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg loss=0.4804860028353604 execution time=0.8617112636566162
	3392/13095 processing file data/processed/train/LJ009-0022 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg loss=0.6009358451479957 execution time=0.818913459777832
	3393/13095 processing file data/processed/train/LJ021-0025 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg loss=0.6074196871589211 execution time=0.6654195785522461
	3394/13095 processing file data/processed/train/LJ029-0019 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg loss=0.45646419525146487 execution time=0.5834240913391113
	3395/13095 processing file data/processed/train/LJ010-0052 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg loss=0.5047349399990506 execution time=0.7163472175598145
	3396/13095 processing file data/processed/train/LJ028-0167 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg loss=0.5539992877415248 execution time=0.5469622611999512
	3397/13095 processing file data/processed/train/LJ046-0178 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg loss=0.5276196935902471 execution time=0.9037868976593018
	3398/13095 processing file data/processed/train/LJ036-0003 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg loss=0.5141079425811768 execution time=0.2747025489807129
	3399/13095 processing file data/processed/train/LJ016-0374 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 avg loss=0.47609440485636395 execution time=0.9370806217193604
	Synthesizing devset
		1/5 processing file data/processed/dev/LJ002-0290 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 execution time=1.0828371047973633
		2/5 processing file data/processed/dev/LJ001-0104 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 execution time=0.467318058013916
		3/5 processing file data/processed/dev/LJ005-0117 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 execution time=1.244554042816162
		4/5 processing file data/processed/dev/LJ001-0092 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 execution time=1.6375510692596436
		5/5 processing file data/processed/dev/LJ001-0009 5 10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95 100 execution time=1.547147274017334
```
It generates 10 seconds of speech in under 2. Let me know what you think.

Best,
Tibi

**Comment by tiberiu44** [comment]
Hi @carlfm01 
Core I3, single NVidia GTX1080 8GB. It implements something in the lines of: https://arxiv.org/abs/1808.06719

Hope this helps.

**Comment by tiberiu44** [comment]
It's real-time on a I7 CPU. The I3 is 4 times slower. Also, I have to mention that I obtained nice results using gold-standard data, but end-to-end speech synthesis created a lot of artifacts. This model is not tolerant to noise added by the text encoder which predicts the mel-spectogram. I'm trying something different at the moment and hopefully I will get real-time speech synthesis with high quality.
